### ISSUE TITLE
🎯 Cultural Relevance Index (CRI) Backtesting System Implementation

### ISSUE THREAD

**Issue by holden** [razrfly/cinegraph#81]
# 🎯 Cultural Relevance Index (CRI) Backtesting System Implementation

## 📋 Overview

Implement a comprehensive backtesting system to validate our Cultural Relevance Index (CRI) algorithm against the authoritative "1001 Movies You Must See Before You Die" list. This system will help us optimize our scoring weights and validate that our algorithm can accurately predict cultural relevance using available data sources.

## 🎯 Primary Objective

Build a machine learning backtesting pipeline that can:
1. **Import the complete "1001 Movies" ground truth dataset**
2. **Normalize existing data sources** (Oscar nominations, OMDb ratings, box office data)
3. **Train and validate CRI scoring weights** to maximize overlap with expert consensus
4. **Measure precision and recall** against the authoritative list
5. **Provide framework for adding new data sources** incrementally

### Available Data Sources
- ✅ **Academy Awards**: 947 nominations across 368 movies (2016-2024)
- ✅ **OMDb Data**: 139 movies with comprehensive metadata
  - IMDb ratings, Metacritic scores, Rotten Tomatoes scores
  - Box office data, awards summaries
  - Genre, runtime, release date information
- ✅ **TMDb Data**: Comprehensive movie database with 1,227+ movies
  - Vote averages, popularity scores, vote counts
  - Cast, crew, keywords, production companies
  - Release dates, budgets, revenues

### Data Quality Metrics
- **Oscar Data**: 100% success rate (1,227/1,227 jobs completed)
- **OMDb Coverage**: 139/361 Oscar-nominated movies have OMDb enrichment
- **External Ratings**: 8 different rating types from multiple sources

### Dataset Sources Identified
1. **IMDb Comprehensive List**: 1,260 movies from all editions (2003-2024)
   - URL: https://www.imdb.com/list/ls024863935/
   - Includes movies from 1902-2023 across all book editions

2. **Fandom Wiki**: Complete historical tracking
   - URL: https://1001films.fandom.com/wiki/The_List  
   - Tracks additions/removals across editions

3. **Letterboxd List**: 1,245 movies from 2003-2021 editions
   - URL: https://letterboxd.com/peterstanley/list/1001-movies-you-must-see-before-you-die/

### Data Extraction Strategy
- **Primary**: Web scraping IMDb list for IMDb IDs and basic metadata
- **Secondary**: Cross-reference with Fandom wiki for historical changes
- **Validation**: Match against IMDb official datasets for data quality

### Phase 1: Ground Truth Data Import 
**Timeline: Week 1-2**

#### Tasks:
- [ ] **Create "1001 Movies" scraper** 
  - Extract IMDb IDs, titles, years from IMDb list
  - Handle pagination and rate limiting
  - Store in `canonical_movies` table with source attribution

- [ ] **Import movies into our database**
  - Use existing TMDb pipeline to fetch comprehensive data
  - Queue through TMDbDetailsWorker → OMDbEnrichmentWorker
  - Track import success/failure rates

- [ ] **Create canonical movie tracking**
  ```sql
  CREATE TABLE canonical_movies (
    id bigserial PRIMARY KEY,
    movie_id bigint REFERENCES movies(id),
    imdb_id varchar(20) NOT NULL,
    source varchar(50) NOT NULL, -- '1001_movies_2024', etc.
    added_in_edition varchar(20),
    removed_in_edition varchar(20),
    canonical_rank integer,
    inserted_at timestamp DEFAULT now()
  );
  ```

#### Success Criteria:
- ✅ 1,200+ canonical movies imported with IMDb IDs
- ✅ 90%+ successful TMDb/OMDb enrichment
- ✅ Complete audit trail of source attributions

### Phase 2: Data Normalization & Feature Engineering
**Timeline: Week 3-4**

#### Oscar Awards Normalization
```elixir
# Scoring functions to implement
defmodule Cinegraph.CRI.Features.Awards do
  @doc "Convert Oscar data to normalized scores"
  def oscar_score(movie) do
    nominations = get_oscar_nominations(movie.id)
    wins = Enum.count(nominations, & &1.won)
    nominees = Enum.count(nominations, & !&1.won)
    
    # Weighted scoring: wins worth more than nominations
    # Major categories (Best Picture, Director, Actor) worth more
    base_score = (wins * 1.0) + (nominees * 0.3)
    
    major_category_bonus = 
      nominations
      |> Enum.filter(& &1.category.name in ["Best Picture", "Directing", "Actor in a Leading Role"])
      |> Enum.count()
      |> Kernel.*(0.5)
    
    normalize_to_0_1(base_score + major_category_bonus)
  end
end
```

#### OMDb Data Normalization
```elixir
defmodule Cinegraph.CRI.Features.Ratings do
  @doc "Normalize ratings from different sources to 0-1 scale"
  def normalized_ratings(movie) do
    %{
      imdb_rating: normalize_imdb_rating(movie.omdb_data["imdbRating"]),
      metacritic_score: normalize_metacritic(movie.omdb_data["Metascore"]),
      rotten_tomatoes: extract_rt_score(movie.omdb_data["Ratings"]),
      awards_summary: parse_awards_summary(movie.omdb_data["Awards"])
    }
  end
  
  @doc "Extract wins/nominations from OMDb Awards string"
  def parse_awards_summary(awards_string) do
    # Parse "Won 3 Oscars. 136 wins & 344 nominations total"
    # Extract oscar_wins, total_wins, total_nominations
    # Convert to normalized cultural impact score
  end
end
```

#### Box Office & Commercial Success
```elixir
defmodule Cinegraph.CRI.Features.Commercial do
  @doc "Normalize box office performance relative to era"
  def box_office_score(movie) do
    raw_box_office = parse_box_office(movie.omdb_data["BoxOffice"])
    release_year = extract_year(movie.release_date)
    
    # Adjust for inflation and market size changes
    inflation_adjusted = adjust_for_inflation(raw_box_office, release_year)
    percentile_in_year = get_yearly_percentile(inflation_adjusted, release_year)
    
    normalize_to_0_1(percentile_in_year)
  end
end
```

#### Temporal Relevance Features
```elixir
defmodule Cinegraph.CRI.Features.Temporal do
  @doc "Account for recency bias and lasting power"
  def temporal_relevance(movie) do
    release_year = extract_year(movie.release_date)
    years_since_release = DateTime.utc_now().year - release_year
    
    # Older films that remain relevant get bonus points
    # But very recent films might not have proven staying power
    age_factor = calculate_age_bonus(years_since_release)
    award_longevity = calculate_award_longevity(movie)
    
    %{
      age_factor: age_factor,
      award_longevity: award_longevity,
      combined_temporal: (age_factor + award_longevity) / 2
    }
  end
end
```

### Phase 3: CRI Algorithm Implementation
**Timeline: Week 5-6**

#### Core CRI Calculation Engine
```elixir
defmodule Cinegraph.CRI.Calculator do
  @doc "Calculate Cultural Relevance Index for a movie"
  def calculate_cri(movie, weights \\ default_weights()) do
    features = %{
      oscar_score: Features.Awards.oscar_score(movie),
      critical_consensus: Features.Ratings.critical_consensus(movie),
      commercial_success: Features.Commercial.box_office_score(movie),
      awards_breadth: Features.Awards.total_awards_score(movie),
      temporal_relevance: Features.Temporal.temporal_relevance(movie),
      popularity_metrics: Features.Popularity.tmdb_popularity(movie)
    }
    
    # Weighted combination
    cri_score = 
      (features.oscar_score * weights.oscar_weight) +
      (features.critical_consensus * weights.critical_weight) +
      (features.commercial_success * weights.commercial_weight) +
      (features.awards_breadth * weights.awards_weight) +
      (features.temporal_relevance * weights.temporal_weight) +
      (features.popularity_metrics * weights.popularity_weight)
    
    # Scale to 0-100 and store
    final_score = cri_score * 100
    
    Cultural.upsert_cri_score(movie.id, final_score, features, "1.0")
    
    %{score: final_score, features: features, movie: movie}
  end
  
  defp default_weights do
    %{
      oscar_weight: 0.35,      # Academy Awards heavily weighted
      critical_weight: 0.20,   # Professional critical consensus  
      commercial_weight: 0.15, # Box office success (inflation adjusted)
      awards_weight: 0.15,     # Broader awards recognition
      temporal_weight: 0.10,   # Staying power over time
      popularity_weight: 0.05  # General popularity metrics
    }
  end
end
```

### Phase 4: Backtesting Framework
**Timeline: Week 7-8**

#### Training/Validation Split
```elixir
defmodule Cinegraph.CRI.Backtesting do
  @doc "Split canonical movies for training/validation"
  def create_training_split(canonical_movies, train_ratio \\ 0.8) do
    # Stratified split by decade to ensure temporal representation
    canonical_movies
    |> Enum.group_by(fn movie -> decade(movie.release_date) end)
    |> Enum.flat_map(fn {decade, movies} ->
      split_size = round(length(movies) * train_ratio)
      {train, test} = Enum.split(Enum.shuffle(movies), split_size)
      [{:train, train}, {:test, test}]
    end)
    |> Enum.group_by(&elem(&1, 0), &elem(&1, 1))
    |> Map.new(fn {key, lists} -> {key, List.flatten(lists)} end)
  end
  
  @doc "Optimize weights using training data"  
  def optimize_weights(training_movies, initial_weights \\ default_weights()) do
    # Use grid search or gradient descent to find optimal weights
    # Objective: maximize overlap between top N predicted movies and canonical list
    
    weight_combinations = generate_weight_grid(initial_weights)
    
    best_weights = 
      weight_combinations
      |> Enum.map(fn weights ->
        precision_recall = evaluate_weights(training_movies, weights)
        {weights, precision_recall}
      end)
      |> Enum.max_by(fn {_weights, metrics} -> metrics.f1_score end)
      |> elem(0)
    
    best_weights
  end
  
  @doc "Evaluate algorithm performance"
  def evaluate_weights(movies, weights) do
    # Calculate CRI for all movies using these weights
    scored_movies = 
      movies
      |> Enum.map(&Calculator.calculate_cri(&1, weights))
      |> Enum.sort_by(& &1.score, :desc)
    
    # Take top N predictions (where N = number of canonical movies)
    canonical_count = length(canonical_movies())
    top_predictions = Enum.take(scored_movies, canonical_count)
    
    # Calculate precision/recall against ground truth
    true_positives = count_overlap(top_predictions, canonical_movies())
    
    %{
      precision: true_positives / length(top_predictions),
      recall: true_positives / canonical_count,
      f1_score: calculate_f1(precision, recall),
      top_100_precision: precision_at_k(top_predictions, 100),
      top_500_precision: precision_at_k(top_predictions, 500)
    }
  end
end
```

### Phase 5: Elixir vs Python Analysis Tools
**Timeline: Week 9**

#### Elixir Libraries Assessment

**Recommended Elixir Stack:**
```elixir
# Data manipulation and analysis
{:explorer, "~> 0.8"},      # DataFrame operations (Pandas equivalent)
{:nx, "~> 0.7"},            # Numerical computing (NumPy equivalent)  
{:scholar, "~> 0.3"},       # Machine learning algorithms (scikit-learn equivalent)
{:axon, "~> 0.6"},          # Neural networks (if needed for advanced models)

# Statistical analysis and optimization
{:statistics, "~> 0.6"},    # Statistical functions
{:optimization, "~> 0.1"},  # Optimization algorithms for weight tuning

# Data visualization (if needed)
{:vega_lite, "~> 0.1"},     # Data visualization
{:kino, "~> 0.12"},         # Livebook integration for interactive analysis
```

**Elixir Implementation Approach:**
- Use **Explorer** for data manipulation and feature engineering
- Use **Scholar** for statistical analysis and basic ML algorithms  
- Use **Nx** for numerical computations and matrix operations
- Leverage **GenServer** for concurrent backtesting across parameter spaces
- Use **Oban** for distributed computation of CRI scores

#### Python Integration Option

**If Elixir ecosystem limitations require Python:**
```python
# Export data from Elixir as JSON/CSV
# Python backtesting stack:
import pandas as pd
import numpy as np
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import precision_recall_fscore_support
from sklearn.ensemble import RandomForestRegressor
import backtesting  # backtesting.py library
```

**Hybrid Approach:**
- **Elixir**: Data ingestion, normalization, real-time CRI calculation
- **Python**: Advanced backtesting, hyperparameter optimization, statistical analysis
- **Integration**: JSON API or shared database for data exchange

### Primary KPIs
- **Precision@1000**: % of our top 1000 predictions that are in canonical list
- **Recall**: % of canonical movies we successfully identify in top predictions  
- **F1-Score**: Harmonic mean of precision and recall
- **Ranking Correlation**: Spearman correlation between our scores and canonical ranking

### Target Performance Goals
- **Precision@1000**: >70% (700+ canonical movies in our top 1000)
- **Recall**: >80% (identify 80%+ of canonical movies somewhere in our rankings)
- **F1-Score**: >0.75 
- **Statistical Significance**: p-value <0.05 vs random selection

### Validation Framework
1. **Cross-validation**: 5-fold cross-validation across decades
2. **Temporal validation**: Train on pre-2000 movies, test on post-2000
3. **Holdout validation**: Reserve 20% of canonical movies for final validation
4. **Ablation studies**: Test impact of removing each feature category

### Phase 6+: Additional Data Sources
Once baseline backtesting is established, incrementally add:

1. **Criterion Collection** (prestige art house films)
2. **National Film Registry** (cultural preservation)
3. **Sight & Sound Critics Poll** (critical consensus)
4. **Google Trends data** (contemporary cultural impact)
5. **Letterboxd ratings** (modern film community)
6. **Academic citations** (scholarly recognition)

### Continuous Improvement Process
1. **Add new data source**
2. **Create normalization functions**
3. **Re-run backtesting with updated features**
4. **Compare performance metrics**
5. **Update production weights if improvement >5%**

### Database Schema Updates
```sql
-- Store backtesting results
CREATE TABLE cri_backtests (
  id bigserial PRIMARY KEY,
  version varchar(20) NOT NULL,
  weights jsonb NOT NULL,
  training_precision decimal(5,4),
  training_recall decimal(5,4),
  validation_precision decimal(5,4),
  validation_recall decimal(5,4),
  f1_score decimal(5,4),
  feature_importance jsonb,
  created_at timestamp DEFAULT now()
);

-- Track canonical movie performance
CREATE TABLE movie_predictions (
  id bigserial PRIMARY KEY,
  backtest_id bigint REFERENCES cri_backtests(id),
  movie_id bigint REFERENCES movies(id),
  predicted_score decimal(8,4),
  predicted_rank integer,
  is_canonical boolean,
  canonical_rank integer
);
```

### API Endpoints for Analysis
```elixir
# Add to router.ex
scope "/api/cri", CinegraphWeb do
  get "/backtest/results", CRIController, :backtest_results
  post "/backtest/run", CRIController, :run_backtest
  get "/movie/:id/features", CRIController, :movie_features
  get "/leaderboard", CRIController, :cri_leaderboard
end
```

### Immediate Next Steps (Week 1)
1. **[ ] Create IMDb scraper** for "1001 Movies" list 
2. **[ ] Set up canonical movies database schema**
3. **[ ] Import first 100 canonical movies as proof of concept**
4. **[ ] Verify TMDb/OMDb enrichment pipeline works for canonical movies**

### Development Priorities
1. **Data Foundation**: Complete canonical movie import (90% priority)
2. **Feature Engineering**: Normalize existing data sources (85% priority)  
3. **Backtesting Framework**: Build validation pipeline (80% priority)
4. **Algorithm Optimization**: Weight tuning and model selection (75% priority)

### Technical Risks
- **Data Quality**: Some canonical movies may not have sufficient data
  - *Mitigation*: Track data completeness, use imputation strategies
- **Elixir ML Ecosystem**: May lack advanced backtesting features
  - *Mitigation*: Start with Elixir, evaluate Python integration if needed
- **Temporal Bias**: Algorithm may favor recent movies with more data
  - *Mitigation*: Include decade-based stratification in training

### Business Risks  
- **Overfitting**: Algorithm performs well on training but not real cultural relevance
  - *Mitigation*: Use multiple validation approaches, test on holdout data
- **Data Drift**: Movie industry and cultural patterns change over time
  - *Mitigation*: Implement continuous retraining and model monitoring

---

## 🎬 Expected Outcome

By the end of this implementation, CineGraph will have:
- ✅ **1,200+ canonical movies** imported with comprehensive data
- ✅ **Validated CRI algorithm** achieving >70% precision against expert consensus  
- ✅ **Production-ready scoring system** that can evaluate any movie's cultural relevance
- ✅ **Backtesting framework** for continuous algorithm improvement
- ✅ **Foundation for expansion** to additional cultural authorities and data sources

This establishes CineGraph as the first **data-driven cultural relevance measurement system** with empirical validation against established film canon.

**Comment by holden** [comment]
SUPERSEDED by #682/#700 — original CRI backtesting issue; backtesting works, ML iteration is the new thread.
